File: python/hyperion_api.py
import requests
from websocket import create_connection
import json


class HyperionAPI:
    api_key = 0
    uri = ""
    protocol = "REST"
    stream_callback = None

    def __init__(self, api_key=0, uri="", protocol="REST", stream_callback=None):
        self.api_key = api_key
        self.uri = uri
        self.protocol = protocol
        self.stream_callback = stream_callback

    def generate(
        self,
        model,
        messages,
        stream=False,
        **kwargs,
    ):
        params = {
            "model": model,
            "messages": messages,
            "stream": stream,
        }
        params.update(kwargs)

        if self.protocol == "REST":
            api_method = "model/generate/"
        else:
            api_method = "model.generate"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def load_model(self, model, stream=False):
        params = {
            "model": model,
            "stream": stream,
        }
        if self.protocol == "REST":
            api_method = "model/load"
        else:
            api_method = "model.load"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def unload_model(self, model):
        params = {"model": model}
        if self.protocol == "REST":
            api_method = "model/unload/"
        else:
            api_method = "model.unload"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def get_loaded_models(self):
        params = None
        if self.protocol == "REST":
            api_method = "model/loaded/"
        else:
            api_method = "model.get_loaded"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def get_cached_models(self):
        params = None
        if self.protocol == "REST":
            api_method = "model/cached/"
        else:
            api_method = "model.get_cached"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def create_embedding(self, model, input):
        params = {
            "model": model,
            "input": input,
        }
        if self.protocol == "REST":
            api_method = "embedding/create/"
        else:
            api_method = "embedding.create"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def tokenize(self, model, text):
        params = {
            "model": model,
            "text": text,
        }

        if self.protocol == "REST":
            api_method = "model/tokenize/"
        else:
            api_method = "model.tokenize"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def detokenize(self, model, tokens):
        params = {
            "model": model,
            "tokens": tokens,
        }
        if self.protocol == "REST":
            api_method = "model/detokenize/"
        else:
            api_method = "model.detokenize"

        try:
            response = self._call(api_method, params)
            return response
        except Exception as error:
            print(f"Hyperion API Error in {api_method}: ", error)
            response = {"success": "false", "error": str(error)}
            return response

    def _call(self, api_method, params=None, stream=False):
        if self.protocol == "REST":
            if params is None:
                response = requests.get(f"http://{self.uri}/{api_method}", json=request)
                response = response.json()
            else:
                request = {"params": params}
                response = requests.post(
                    f"http://{self.uri}/{api_method}", json=request
                )
                response = response.json()
            return response
        elif self.protocol == "WEBSOCKET":
            websocket = create_connection(f"ws://{self.uri}")
            request = {"method": f"{api_method}", "params": params}
            stream = params["stream"]
            websocket.send(json.dumps(request))

            if stream == True and self.stream_callback is not None:
                while True:
                    response = json.loads(websocket.recv())
                    self.stream_callback(response)
                    if response.get("stream_end") == True:
                        break

            # Get the final complete output
            response = websocket.recv()
            response = json.loads(response)
            websocket.close()
            return response
        else:
            return {"success": "false", "error": "Unknown server mode"}

File: javascript/HyperionAPI.js
const WebSocket = require("ws");

class HyperionAPI {
	constructor({ apiKey = 0, uri = "", protocol = "REST", streamCallback = null } = {}) {
		this.apiKey = apiKey;
		this.uri = uri;
		this.protocol = protocol;
		this.streamCallback = streamCallback;
	}

	async generate({
		model = "",
		messages = [],
		stream = false,
		maxNewTokens = null,
		temperature = null,
		doSample = null,
		topP = null,
		topK = null,
		repetitionPenalty = null,
		lengthPenalty = null,
	} = {}) {
		let params = {
			model: model,
			messages: messages,
			stream: stream,
			max_new_tokens: maxNewTokens,
			temperature: temperature,
			do_sample: doSample,
			top_p: topP,
			top_k: topK,
			repetition_penalty: repetitionPenalty,
			length_penalty: lengthPenalty,
		};

		// Filter out parameters with null values
		for (const key in params) {
			if (params[key] === null) {
				delete params[key];
			}
		}

		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/generate/";
		} else {
			apiMethod = "model.generate";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async loadModel({ model = "", stream = false } = {}) {
		let params = {
			model: model,
			stream: stream,
		};

		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/load/";
		} else {
			apiMethod = "model.load";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async unloadModel({ model = "" } = {}) {
		let params = { model: model };

		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/unload/";
		} else {
			apiMethod = "model.unload";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async getLoadedModels() {
		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/loaded/";
		} else {
			apiMethod = "model.loaded";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async getCachedModels() {
		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/cached/";
		} else {
			apiMethod = "model.cached";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async createEmbedding({ model = "", input = "" } = {}) {
		let params = {
			model: model,
			input: input,
		};

		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "embedding/create/";
		} else {
			apiMethod = "embedding.create";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async tokenize({ model = "", text = "" } = {}) {
		let params = {
			model: model,
			text: text,
		};

		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/tokenize/";
		} else {
			apiMethod = "model.tokenize";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async detokenize({ model = "", tokens = [] } = {}) {
		let params = {
			model: model,
			tokens: tokens,
		};

		let apiMethod = "";
		if (this.protocol === "REST") {
			apiMethod = "model/detokenize/";
		} else {
			apiMethod = "model.detokenize";
		}

		try {
			const response = await this._call(apiMethod, params);
			return response;
		} catch (error) {
			console.log("Hyperion API Error in " + apiMethod + ": ", error.message);
			const response = { success: "false", error: error.message };
			return response;
		}
	}

	async _call(apiMethod, params = null) {
		if (this.protocol === "REST") {
			const request = { params: params };
			let response = null;
			if (params !== null) {
				response = await fetch("http://${this.uri}/" + apiMethod, {
					method: "POST",
					headers: {
						"Content-Type": "application/json",
					},
					body: JSON.stringify(request),
				});
			} else {
				response = await fetch(`http://${this.uri}/model/cached/`, {
					method: "GET",
				});
			}
			const jsonResponse = await response.json();
			return jsonResponse;
		} else if (this.protocol === "WEBSOCKET") {
			return new Promise((resolve, reject) => {
				const websocket = new WebSocket(`ws://${this.uri}`);
				const request = { method: apiMethod, params: params };
				const stream = params && "stream" in params ? params.stream : false;
				websocket.onopen = () => {
					websocket.send(JSON.stringify(request));
				};
				websocket.onmessage = (event) => {
					const response = JSON.parse(event.data);
					// Only send streamed tokens to the callback. Do not close the socket
					// until the stream ends and we received the final response.
					if (stream === true && this.streamCallback !== null && response.stream_end != null) {
						this.streamCallback(response);
					}

					// This is the final response, so close the websocket
					// and resolve the promise
					if (response.success == "true" || response.success == "false") {
						resolve(response);
						websocket.close();
					}
				};
				websocket.onerror = (err) => {
					reject(err);
				};
			});
		} else {
			return { success: "false", error: "Unknown server mode" };
		}
	}
}

module.exports = HyperionAPI;
